demos/yielding/README.md

```diff
@@ -0,0 +1,112 @@
+# Yielding
+
+In this demo, we try to show the usage of yielding, which is one characteristic of a multitasking system. 
+
+By calling syscall `yield()`, the thread gives away the utilization of a CPU to another thread, or a specific thread (will be discussed later in another demo).
+
+To achieve thread yielding, we need to introduce the following functions:
+
+* `add_thread_from_worker` (in `src/user/thread.c`)
+* `do_scheduler` (in `src/kernel/thread.c`)
+* `yield` (in `src/user/thread.c`)
+
+## `add_thread_from_worker`
+
+In AMT, we try to maintain all of the threads in an array called `all_threads`.
+
+By calling this function, the AMT will create a new thread instance with the given arguments, initialize the permissions, and put the function in. 
+
+After the creation of the thread, the entity will be put into `all_threads` array and wait for the utilization of the CPU.
+
+## `do_scheduler`
+
+The job of the scheduler is to choose the next thread to make use of the CPU. 
+
+In AMT, the scheduler simply walks through the `all_threads` array to find the next thread.
+
+To run the thread chosen by the scheduler and restore the function call stack of the current thread, we need to save the information of this thread first.
+
+To achieve the goal, the context switch routine written in assembly (`yield.S`) will be executed to find the next task. 
+
+And the task of the thread to be switched will be suspended until the thread has been chosen to run
+
+## `yield`
+
+To yield the utilization, the thread calls the syscall yield and set the kind of yielding as `YIELD_ANY`, which means that we do not specify the thread to acquire the CPU. As a result, the scheduler will help us to choose the next thread to run.
+
+And the current thread will have to wait until next time a thread had yielded the utilization to it.
+
+## The progress of yielding demo
+
+At the beginning of the demo, to observe the behaviour of the scheduler, we use `set_kernel_config` to allow the scheduler to print out the log. 
+
+After setting up the config, we use `add_thread_from_worker` to create a new thread running the `thread_worker`. 
+
+When the initialization is complete, the first thread starts to run into an endless loop and repeatedly yield to others.
```

Comment:
"the first thread (`setup`)"

I confused myself reading this because I thought we had two threads in **addition** to the setup thread. Would be good to note that just for people like me who aren't paying attention :)